src/loop_rerolling.py

```diff
@@ -0,0 +1,116 @@
+from typing import List
+
+from .flow_graph import (
+    BasicNode,
+    ConditionalNode,
+    FlowGraph,
+    Node,
+    compute_relations,
+)
+from .parse_instruction import Instruction
+
+
+def replace_node_references(
+    flow_graph: FlowGraph, replace_this: Node, with_this: Node
+) -> None:
+    for node_to_modify in flow_graph.nodes:
+        node_to_modify.replace_any_children(replace_this, with_this)
+
+
+def remove_node(flow_graph: FlowGraph, to_delete: Node, new_child: Node) -> None:
+    flow_graph.nodes.remove(to_delete)
+    replace_node_references(flow_graph, to_delete, new_child)
+
+
+def replace_node(flow_graph: FlowGraph, replace_this: Node, with_this: Node) -> None:
+    replacement_index = flow_graph.nodes.index(replace_this)
+    flow_graph.nodes[replacement_index] = with_this
+    replace_node_references(flow_graph, replace_this, with_this)
+
+
+def reroll_loop(flow_graph: FlowGraph, start: ConditionalNode) -> bool:
+    node_1 = start.fallthrough_edge
+    node_7 = start.conditional_edge
+
+    if not isinstance(node_1, ConditionalNode):
+        return False
+    node_2 = node_1.fallthrough_edge
+    node_5 = node_1.conditional_edge
+
+    if not isinstance(node_2, BasicNode):
```

Comment:
I think we should also check that these nodes have the parents we expect. I feel like this pass has a large potential for misdetection and changing the asm into something non-equivalent, so we should be as careful as we can in the pattern-matching.